### PR TITLE
OpenSSL 3.0: PKCS11 signers

### PR DESCRIPTION
**Pull request by fredmorcos**
### Short description
Some cleanup and support openssl 3 in `pkcs11signers.cc`.

### Checklist
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/pkcs11signers.cc [hunk 1]
```diff
@@ -211,11 +213,11 @@ class P11KitAttribute {
 
 class Pkcs11Slot {
   private:
-    bool d_logged_in;
+    bool d_logged_in{};
```
Comment (rgacogne) [inline pdns/pkcs11signers.cc:216]: I wouldn't mind setting `d_logged_in` and `d_err` to `false` explicitly. The default-initialization of integral types should be zero-initialization, but I always have to think twice (at least!) to be sure.

File: pdns/pkcs11signers.cc [hunk 2]
```diff
@@ -42,8 +44,8 @@ in it. you need to use softhsm tools to manage this all.
 static CK_FUNCTION_LIST** p11_modules;
 #endif
 
-#define ECDSA256_PARAMS "\x06\x08\x2a\x86\x48\xce\x3d\x03\x01\x07"
-#define ECDSA384_PARAMS "\x06\x05\x2b\x81\x04\x00\x22"
+constexpr const char* ECDSA256_PARAMS{"\x06\x08\x2a\x86\x48\xce\x3d\x03\x01\x07"};
+constexpr const char* ECDSA384_PARAMS{"\x06\x05\x2b\x81\x04\x00\x22"};
```
Comment (rgacogne) [inline pdns/pkcs11signers.cc:48]:
```suggestion
static constexpr const char* ECDSA256_PARAMS{"\x06\x08\x2a\x86\x48\xce\x3d\x03\x01\x07"};
static constexpr const char* ECDSA384_PARAMS{"\x06\x05\x2b\x81\x04\x00\x22"};
```

**Review by rgacogne — APPROVED** [review]

**Review by rgacogne — APPROVED** [review]

**Comment by fredmorcos** [comment]
Switched to draft because I want to test it locally first since it's non-tested on CI for the time being.

**Comment by fredmorcos** [comment]
Re-opening for review:
- I tested it locally and asan complained about a buffer overflow from an unrelated part of the signer. I fixed this.
- Tested on master with the overflow fix and it works.
- Tested on this branch with the overflow fix and it works.
- Rebased on master.

I'm including the overflow fix here instead of opening a new PR for a 3 character long fix.

File: pdns/pkcs11signers.cc [hunk 3]
```diff
@@ -416,7 +437,7 @@ class Pkcs11Token {
 
       size_t k;
       auto pubAttr = std::make_unique<CK_ATTRIBUTE[]>(pubAttributes.size());
-      auto privAttr = std::make_unique<CK_ATTRIBUTE[]>(pubAttributes.size());
+      auto privAttr = std::make_unique<CK_ATTRIBUTE[]>(privAttributes.size());
```
Comment (rgacogne) [inline pdns/pkcs11signers.cc:440]: Wow, nice catch! This code should really use vectors instead..

**Review by rgacogne — APPROVED** [review]
Looks good to me, but this file really needs a huge refactoring..

**Comment by fredmorcos** [comment]
> Looks good to me, but this file really needs a huge refactoring..

https://github.com/PowerDNS/pdns/pull/11151

**Pull request merged**